### ISSUE TITLE
DirectDraw Exclusive Mode compatibility and potential Feature Request to be used from other process

### ISSUE THREAD

**Issue by Kalmat** [rafzi/hacklib#5]
Hi! Very nice and interesting project. Thank you for sharing!

I'm developing a cross-plaform System Monitor (on Python3, just for fun and personal use... or whoever else wants to use it!) which graphically shows key system values (CPU usage and temp, Disk usage, Memory usage, GPU usage, temp and fan speed). It also shows an FPS counter to check in-game performance. The problem is that it doesn't stays on top when working with games using DirectDraw Exclusive Mode, making it useless. After searching a lot I have realized that it can only be done by hooking D3D dll, but this can not be done with Python. Your WindowOverlay is very promissing, so I have two questions: 

- Does it work with DirectDraw Exclusive Mode applications (reading the docs, I assume it does)?
- Is it possible to somehow address the overlay from a separate process so you can draw some info on it (just text could be enough) and, if so, how?

The main problem is that I am totally unskilled in C, my apologies for that. Perhaps we could work on a Python wrapper for the whole thing if you think it's interesting. I believe I can do or help you to do that (as a first idea, loading the C code built as DLL's and exposing its API in a python module).

Thank you in advance for your time and patience!

![Captura de pantalla 2023-01-11 122150](https://user-images.githubusercontent.com/39727793/211799361-e04c6bd4-30f3-4661-a6d6-f8e6ff2558ea.png)
![Captura de pantalla 2023-01-11 122256](https://user-images.githubusercontent.com/39727793/211799357-e9a3762a-c4fb-4e99-9807-dda7cbaae0f6.png)
![Captura de pantalla 2023-01-11 122225](https://user-images.githubusercontent.com/39727793/211799359-574111aa-9b16-40ff-b6fb-464a59a4fad4.png)

**Comment by rafzi** [comment]
Thank for your interest!

Exclusive mode by definition means that only a single graphics pipeline is rendering to the screen. The overlay contained in this library is a separate window, probably similar to what you have built. Therefore, it can not be used the way you imagined.

The only way to achieve this would be to reuse the rendering pipeline of the target process. You would hook a rendering function called every frame, for example EndScene for D3D9, and insert your additional drawing calls. Out-of-the-box, hacklib only really supports this for D3D9 and OpenGL. For the other D3D versions and DirectDraw, a Drawer class would be missing. The DeviceFetcher would also require updates for other D3D versions, DirectDraw and OpenGL.

There is no built-in mechanism for communicating with another process. You'd need some form of IPC:

https://en.wikipedia.org/wiki/Inter-process_communication
https://stackoverflow.com/questions/6915191/simple-ipc-between-c-and-python-cross-platform

Building a Python wrapper sounds like an interesting project. However, I believe this would be a layer on top of the existing code, without any changes required to this library itself. One concern would be performance. For example, you wouldn't want to do 1000s of draw calls every frame from Python wrapper to target process with individual IPC messages.

**Comment by Kalmat** [comment]
Thanks for your prompt reply!

So sad to hear that! HAHAHAHA! I thought I had found the solution! I will keep on searching (I know nothing about C nor DirectX, my bad). Anyway, your project seems really useful and complete. 

Regarding the Python wrapper I was thinking on wrapping the whole project, not just the overlay. It seems useful for Python projects too.

As a curiosity and only if it can be interesting for future reference, while searching for a solution about the overlay, I found [this](https://github.com/Andrey1994/game_overlay_sdk). It seems to do exactly what you said it is necessary to do in order to draw an always-on-top overlay, and it's already wrapped with python (not sure what it uses as IPC, I think it's a file and a queue to avoid the performance issues you were mentioning). The problem is that it only allows text, does not support modifying the overlay in any way (size, position, font, colors, ...) and it's now unsupported (the author told me he was considering to completely rebuild it and possibly make a non-open source solution).

Thank you again for your time. Keep pushing!
